### PR TITLE
Action entity fetch 9547355

### PR DESCRIPTION
**Pull request by diarmy**
Entity fetch by property action - see https://www.drupal.org/node/2317233

@diarmy

### CODE REVIEW

File: src/Plugin/Action/EntityPropertyFetch.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    $property_value = $this->getContextValue('value');
+    $limit = $this->getContextValue('limit');
+
+    $storage = $this->entityManager->getStorage($entity_type);
+
+    // When retrieving entities, if $limit is not set there is no need to use query object directly
+    $entities = array();
+    if (is_null($limit)) {
+      $entities = $storage->loadByProperties(array($entity_property => $property_value));
+    } else {
```
Comment (klausi) [inline src/Plugin/Action/EntityPropertyFetch.php]: Else should be on the next line, see https://www.drupal.org/coding-standards#controlstruct

**Comment by klausi** [comment]
Could you add a link to the issue on drupal.org to the summary?

File: src/Plugin/Action/EntityPropertyFetch.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  public function execute() {
+    // Retrieve context values for action
+    $entity_type = $this->getContextValue('type');
+    $entity_property = $this->getContextValue('property');
+    $property_value = $this->getContextValue('value');
+    $limit = $this->getContextValue('limit');
+
+    $storage = $this->entityManager->getStorage($entity_type);
+
+    // When retrieving entities, if $limit is not set there is no need to use query object directly
```
Comment (klausi) [inline src/Plugin/Action/EntityPropertyFetch.php]: Comments must end with a dot.

File: src/Plugin/Action/EntityPropertyFetch.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    } else {
+      $query = $storage->getQuery();
+      $entity_ids = $query
+        ->condition($entity_property, $property_value, '=')
+        ->range(0, $limit)
+        ->execute();
+      $entities = $storage->loadMultiple($entity_ids);
+    }
+
+    // Set provided value
```
Comment (klausi) [inline src/Plugin/Action/EntityPropertyFetch.php]: Comments must end with a dot.

File: tests/src/Integration/Action/EntityPropertyFetchTest.php [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        $this->moduleHandler,
+        $this->cacheBackend,
+        $this->getMock('Drupal\Core\Language\LanguageManagerInterface'),
+        $this->getStringTranslationStub(),
+        $this->getClassResolverStub(),
+        $this->typedDataManager,
+        $this->getMock('Drupal\Core\KeyValueStore\KeyValueStoreInterface'),
+        $this->getMock('Symfony\Component\EventDispatcher\EventDispatcherInterface')
+      ])
+      ->getMock();
```
Comment (klausi) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php:46]: Why do you have to initialize the entity manager again? This is already don in the parent class? Please add a comment.

File: tests/src/Integration/Action/EntityPropertyFetchTest.php [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    $this->assertEquals('Fetch entities by property', $this->action->summary());
+  }
+
+  /**
+   * Tests action execution when no value for limit is provided.
+   *
+   * @covers ::execute()
+   */
+  public function testActionExecutionWithNoLimit() {
+    // Create variables for action context values
```
Comment (klausi) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php]: Missing dot again, please check all your comments.

File: tests/src/Integration/Action/EntityPropertyFetchTest.php [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    $entity_type = 'entity_test';
+    $property_name = 'test_property';
+    $property_value = 'llama';
+
+    // Create an array of dummy entities
+    $entities = array();
```
Comment (klausi) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php]: we use the PHP short array syntax in Rules.

File: src/Plugin/Action/EntityPropertyFetch.php [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+ *     ),
+ *     "limit" = @ContextDefinition("integer",
+ *       label = @Translation("Limit"),
+ *       description = @Translation("Limit the maximum number of fetched entities."),
+ *       required = FALSE,
+ *     ),
+ *   },
+ *   provides = {
+ *      "entity_fetched" = @ContextDefinition("list",
+ *        label = @Translation("Fetched entity"),
```
Comment (klausi) [inline src/Plugin/Action/EntityPropertyFetch.php:43]: this should be of type "entity" and should have the property multiple = TRUE

File: src/Plugin/Action/EntityPropertyFetch.php [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+      $entity_ids = $query
+        ->condition($entity_property, $property_value, '=')
+        ->range(0, $limit)
+        ->execute();
+      $entities = $storage->loadMultiple($entity_ids);
+    }
+
+    // Set provided value
+    $this->setProvidedValue('entity_fetched', $entities);
+  }
```
Comment (klausi) [inline src/Plugin/Action/EntityPropertyFetch.php:127]: So we should also do a $this->getContextDefinition()->setDataType("entity:$entity_type"); here and add a test case for that. The provided context after the action execution should have the correct type.

File: tests/src/Integration/Action/EntityPropertyFetchTest.php (hunk 4 above)
Comment (milkovsky) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php:46]: imho it is fine. I also did it in https://www.drupal.org/node/2317199. entityManager from RulesEntityIntegrationTestBase does not mock getStorage method

Comment (klausi) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php:46]: Ok, so that reason should be in the comment.

Comment (diarmy) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php:46]: Thanks, @milkovsky and @klausi for the input. I'll update accordingly.

File: tests/src/Integration/Action/EntityPropertyFetchTest.php [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    $this->action->setContextValue('type', $entity_type)
+        ->setContextValue('property', $property_name)
+        ->setContextValue('value', $property_value)
+        ->execute();
+
+    // Test that the provided context has the correct entity type.
+    $this->assertEquals('entity:' . $entity_type, $this->action->getProvidedDefinition('entity_fetched')->getDataType());
+  }
+
+}
```
Comment (milkovsky) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php]: here should be an empty line

Comment (diarmy) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php]: @milkovsky, thanks! Line now added in with a couple of other formatting corrections. Do you know why it's failing Travis?

**Comment by klausi** [comment]
grml, the run-test.sh core patch is broken again.

File: tests/src/Integration/Action/EntityPropertyFetchTest.php (hunk 9 above)
Comment (milkovsky) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php]: I guess there could be an issue with travis configuration. I don't know what is in composer.json and why it wants to patch drupal core. Try to ask at d.o. issue

File: src/Plugin/Action/EntityPropertyFetch.php [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    $entity_property = $this->getContextValue('property');
+    $property_value = $this->getContextValue('value');
+    $limit = $this->getContextValue('limit');
+
+    $storage = $this->entityManager->getStorage($entity_type);
+
+    // When retrieving entities, if $limit is not set there is no need to use query object directly.
+    $entities = [];
+    if (is_null($limit)) {
+      $entities = $storage->loadByProperties(array($entity_property => $property_value));
```
Comment (klausi) [inline src/Plugin/Action/EntityPropertyFetch.php]: This should use the short array syntax.

File: tests/src/Integration/Action/EntityPropertyFetchTest.php [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    for ($i = 0; $i < 2; $i++) {
+      $entity = $this->getMock('Drupal\Core\Entity\EntityInterface');
+      $entities[] = $entity;
+    }
+
+    // Create dummy entity storage object.
+    $entityStorage = $this->getMock('Drupal\Core\Entity\EntityStorageInterface');
+    $entityStorage->expects($this->once())
+      ->method('loadByProperties')
+      ->with(array($property_name => $property_value))
```
Comment (klausi) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php]: Same here, should use short array syntax.

File: tests/src/Integration/Action/EntityPropertyFetchTest.php [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+   * @covers ::execute()
+   */
+  public function testActionExecutionWithLimit() {
+    $entity_type = 'entity_test';
+    $property_name = 'test_property';
+    $property_value = 'llama';
+    $limit = 2;
+
+    // Create an array of dummy entities.
+    $entities = array();
```
Comment (klausi) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php]: should use short array syntax.

File: tests/src/Integration/Action/EntityPropertyFetchTest.php [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+    for ($i = 0; $i < 2; $i++) {
+      $entity = $this->getMock('Drupal\Core\Entity\EntityInterface');
+      $entities[] = $entity;
+    }
+
+    // Create dummy entity storage object.
+    $entityStorage = $this->getMock('Drupal\Core\Entity\EntityStorageInterface');
+    $entityStorage->expects($this->once())
+      ->method('loadByProperties')
+      ->with(array($property_name => $property_value))
```
Comment (klausi) [inline tests/src/Integration/Action/EntityPropertyFetchTest.php]: should use short array syntax.

**Comment by klausi** [comment]
continued in #157

**Pull request closed without merging**